excalidraw/excalidraw
# feat: resave to png/svg with metadata if you loaded your scene from a png/svg file

This is a cleaned-up version of #3516, and fixes #2261. It doesn't attempt anywhere near the number of refactors that I outlined in that PR, because I realised that many of the functions I wanted to change were already part of the public API.

* It is based on top of #3580 (which is required to save scale as appState.exportScale), so you can ignore everything before `add saveType to AppState` when reviewing, if you have already reviewed that PR.
* it adds appState.saveType (png or svg for images with embedded scene, or null for the default .excalidraw json format)
* It adds a new `resaveAsImageWithScene()` helper, as shown in the diagram below, which calls out to the exportScene machinery when you cmd+S.

![plan-2](https://user-images.githubusercontent.com/254647/119684323-53e89a00-be3c-11eb-9bd6-9154f47660ea.png)


Steps to test:

* Download the above diagram (which is a png file that contains an excalidraw scene).
* In an excalidraw from this branch/pr build:
  * open up the png image that you just downloaded
  * make an edit
  * cmd+s to save (on macos. Other operating systems are available)
  * verify that your changes have been persisted to the png image.

Things that have not changed:

* If you export as an image with embedded scene, it will not set fileHandle or saveType, so:
  * cmd+s will still continue to save to your .exalidraw file rather than your image file.
  * If you want cmd+s to resave to the image, you will need to open up the image file after exporting.
